rust/projects/file-io/src/kv.rs

```diff
@@ -15,22 +20,43 @@ use crate::Result;
 /// assert_eq!(val, Some("value".to_owned()));
 /// ```
 pub struct KvStore {
+    // directory for wal and other data
+    #[allow(dead_code)]
+    path: PathBuf,
+    wal_writer: BufWriter<File>,
     map: HashMap<String, String>,
 }
 
 impl KvStore {
-    /// Creates a `KvStore`.
-    pub fn new() -> KvStore {
-        KvStore {
-            map: HashMap::new(),
-        }
+    /// Opens a `KvStore` with the given path.
+    ///
+    /// This will create a new directory if the given one does not exist.
+    pub fn open(path: impl Into<PathBuf>) -> Result<KvStore> {
+        let path = path.into();
+        create_dir_all(&path)?;
+
+        let mut wal_path = path.clone();
+        wal_path.push("wal.log");
+        let wal = OpenOptions::new()
+            .create(true)
+            .read(true)
+            .append(true)
+            .open(&wal_path)?;
+
+        Ok(KvStore {
+            path,
+            wal_writer: BufWriter::new(wal),
+            map: KvStore::load_from_wal(&wal_path)?,
+        })
     }
 
     /// Sets the value of a string key to a string.
     ///
     /// If the key already exists, the previous value will be overwritten.
     pub fn set(&mut self, key: String, value: String) -> Result<()> {
-        self.map.insert(key, value);
+        let cmd = SetCommand::new(key, value);
+        to_writer(&mut self.wal_writer, &cmd)?;
```

Comment:
It took me a few seconds to figure out that `to_writer` isn't in this crate, but from serde_json.

It's common in Rust to import types, but to call functions through their module name, like `serde_json::to_writer`. Will you do that here for clarity?